File: src/policy/url-guard.ts
/**
 * URL guard — blocks SSRF-prone, local, and dangerous-scheme URLs in tool arguments.
 */
import { walkStringLeaves } from './arg-leaf-walker.js';
import { stripZeroWidthCharacters } from '../utils/payload-normalizer.js';
import { isTrustedDomainSquat } from '../utils/registrable-domain.js';

/** URL guard normalization — percent-decode + NFKC; avoids base64/hex blob decode on hostnames. */
function normalizeUrlInput(raw: string): string {
  let current = stripZeroWidthCharacters(raw.trim()).normalize('NFKC');
  for (let i = 0; i < 5; i++) {
    const before = current;
    current = current.replace(/%([0-9A-Fa-f]{2})/g, (_m, hex) => {
      try {
        return String.fromCharCode(parseInt(hex, 16));
      } catch {
        return _m;
      }
    });
    if (current === before) break;
  }
  return current;
}

const URL_ARG_FIELDS = new Set(['url', 'href', 'target', 'webhook', 'callback', 'link']);

/** Freetext fields that may embed http(s) URLs (SSRF in message/body, not only dedicated url keys). */
const FREETEXT_URL_ARG_FIELDS = new Set([
  ...URL_ARG_FIELDS,
  'message',
  'query',
  'body',
  'content',
  'text',
  'prompt',
]);

const PUPPETEER_TOOLS = new Set(['puppeteer_navigate', 'puppeteer_screenshot']);

/** Public documentation hosts allowed for benign corpus (ssrf-025/026). */
const DOCUMENTATION_HOST_ALLOWLIST = new Set([
  'example.com',
  'www.example.com',
  'docs.example.com',
]);

/** Legitimate schema hosts only — blocks evil.schema.org style squatting. */
const ALLOWED_SPEC_SCHEMA_HOSTS = new Set([
  'schema.org',
  'www.schema.org',
  'json-schema.org',
  'www.json-schema.org',
]);

function isSpecDomainSquat(host: string): boolean {
  const h = host.toLowerCase();
  if (ALLOWED_SPEC_SCHEMA_HOSTS.has(h)) return false;
  if (/^[\w-]+\.schema\.org$/i.test(h)) return true;
  if (/^[\w-]+\.json-schema\.org$/i.test(h)) return true;
  return false;
}

const BLOCKED_SCHEMES = new Set(['file', 'javascript', 'data', 'vbscript', 'about']);

const LOCALHOST_NAMES = new Set([
  'localhost',
  'localhost.localdomain',
  'metadata',
  'metadata.google.internal',
  'metadata.google',
  'kubernetes.default.svc',
]);

/** Link-local / cloud metadata endpoints (SSRF). */
const METADATA_IPV4 = /^169\.254\./;

const PRIVATE_IPV4_OCTETS = [
  /^127\./,
  /^10\./,
  /^192\.168\./,
  /^172\.(?:1[6-9]|2\d|3[01])\./,
  /^0\./,
  /^100\.(?:6[4-9]|[7-9]\d|1[01]\d|12[0-7])\./, // CGNAT / shared
  /^169\.254\./,
];

const HTTP_URL_IN_TEXT = /https?:\/\/[^\s"'<>]+/gi;

/** Admin / control-plane paths targeted by browser automation attacks. */
const SENSITIVE_ADMIN_PATH_PATTERNS: RegExp[] = [
  /^\/admin(?:\/|$)/i,
  /^\/wp-admin(?:\/|$)/i,
  /^\/wp-login\.php$/i,
  /^\/dashboard(?:\/|$)/i,
  /^\/internal(?:\/|$)/i,
  /^\/management(?:\/|$)/i,
  /^\/console(?:\/|$)/i,
  /^\/settings(?:\/|$)/i,
  /^\/actuator(?:\/|$)/i,
  /^\/grafana(?:\/|$)/i,
  /^\/phpmyadmin(?:\/|$)/i,
  /^\/\.env$/i,
  /^\/config(?:\/|$)/i,
];

function isPrivateOrLocalIpv4(host: string): boolean {
  return PRIVATE_IPV4_OCTETS.some((re) => re.test(host));
}

function isDecimalIpHost(host: string): boolean {
  if (!/^\d{1,10}$/.test(host)) return false;
  const n = Number(host);
  if (!Number.isInteger(n) || n < 0 || n > 0xffffffff) return false;
  return host === String(n);
}

function decimalIpToDotted(decimal: number): string {
  return [
    (decimal >>> 24) & 0xff,
    (decimal >>> 16) & 0xff,
    (decimal >>> 8) & 0xff,
    decimal & 0xff,
  ].join('.');
}

function isPrivateOrLocalIpv6(host: string): boolean {
  const h = host.toLowerCase();
  if (h === '::1' || h === '::') return true;
  if (h.startsWith('fe80:')) return true; // link-local
  if (h.startsWith('fc') || h.startsWith('fd')) return true; // ULA
  if (h.startsWith('::ffff:')) {
    const v4 = h.slice('::ffff:'.length);
    if (/^[\da-f.]+$/i.test(v4)) return isPrivateOrLocalIpv4(v4);
  }
  return false;
}

function parseUrlCandidate(raw: string): URL | null {
  const trimmed = raw.trim();
  if (!trimmed) return null;
  try {
    const withScheme = /^[a-z][a-z0-9+.-]*:/i.test(trimmed) ? trimmed : `http://${trimmed}`;
    return new URL(withScheme);
  } catch {
    return null;
  }
}

function hostnameFromParsed(url: URL): string {
  if (url.hostname.startsWith('[') && url.hostname.endsWith(']')) {
    return url.hostname.slice(1, -1);
  }
  return url.hostname;
}

/** Normalize dotted-quad octets (e.g. 0177.0.0.1 → 127.0.0.1, 0x7f.0.0.1). */
function normalizeDottedHost(host: string): string {
  if (!/^[\d.x]+$/i.test(host) || !host.includes('.')) return host;
  const parts = host.split('.');
  if (parts.length !== 4) return host;
  const normalized: number[] = [];
  for (const part of parts) {
    if (/^0x[0-9a-f]+$/i.test(part)) {
      normalized.push(parseInt(part, 16) & 0xff);
    } else if (/^0[0-7]+$/.test(part) && part.length > 1) {
      normalized.push(parseInt(part, 8) & 0xff);
    } else {
      const n = parseInt(part, 10);
      if (!Number.isFinite(n) || n < 0 || n > 255) return host;
      normalized.push(n);
    }
  }
  return normalized.join('.');
}

export function isDangerousUrl(raw: string): { block: boolean; reason?: string } {
  const trimmed = normalizeUrlInput(raw);
  if (!trimmed) return { block: false };

  const parsed = parseUrlCandidate(trimmed);
  if (!parsed) {
    if (/^(?:file|javascript|data|vbscript):/i.test(trimmed)) {
      return { block: true, reason: `Blocked URL scheme: ${trimmed.slice(0, 32)}` };
    }
    return { block: false };
  }

  const scheme = parsed.protocol.replace(/:$/, '').toLowerCase();
  if (BLOCKED_SCHEMES.has(scheme)) {
    return { block: true, reason: `Blocked URL scheme (${scheme})` };
  }

  let host = hostnameFromParsed(parsed).toLowerCase();
  if (/^[\d.x]+$/i.test(host)) {
    host = normalizeDottedHost(host);
  }

  if (LOCALHOST_NAMES.has(host) || host.endsWith('.localhost')) {
    return { block: true, reason: `Blocked localhost/metadata host: ${host}` };
  }

  if (isDecimalIpHost(host)) {
    const dotted = decimalIpToDotted(Number(host));
    if (isPrivateOrLocalIpv4(dotted) || METADATA_IPV4.test(dotted)) {
      return { block: true, reason: `Blocked decimal IP (maps to ${dotted})` };
    }
  }

  if (/^[\d.]+$/.test(host)) {
    if (isPrivateOrLocalIpv4(host) || METADATA_IPV4.test(host)) {
      return { block: true, reason: `Blocked private/metadata IP: ${host}` };
    }
  }

  if (host.includes(':') && isPrivateOrLocalIpv6(host)) {
    return { block: true, reason: `Blocked local/private IPv6: ${host}` };
  }

  if (METADATA_IPV4.test(host)) {
    return { block: true, reason: `Blocked metadata IP: ${host}` };
  }

  if (DOCUMENTATION_HOST_ALLOWLIST.has(host)) {
    return { block: false };
  }

  return { block: false };
}

export function extractUrlArgumentValues(
  args: Record<string, unknown> | undefined,
  toolName?: string,
): string[] {
  const values: string[] = [];
  if (!args) return values;

  const scanAllLeaves = toolName && PUPPETEER_TOOLS.has(toolName);

  for (const [key, val] of Object.entries(args)) {
    const keyLower = key.toLowerCase();
    if (FREETEXT_URL_ARG_FIELDS.has(keyLower)) {
      if (typeof val === 'string') values.push(val);
    } else if (scanAllLeaves && typeof val === 'string') {
      values.push(val);
    }
  }

  return values;
}

export function extractHttpUrlsFromLeaves(obj: unknown): string[] {
  const urls: string[] = [];
  for (const { path, value } of walkStringLeaves(obj)) {
    const key = path.split(/[.[\]]/).filter(Boolean).pop()?.toLowerCase() ?? '';
    if (FREETEXT_URL_ARG_FIELDS.has(key)) {
      urls.push(value);
    }
    for (const m of value.matchAll(HTTP_URL_IN_TEXT)) {
      urls.push(m[0]);
    }
  }
  return urls;
}

export interface UrlGuardResult {
  block: boolean;
  reason?: string;
}

function expandUrlCandidates(urls: string[]): string[] {
  const expanded: string[] = [];
  for (const raw of urls) {
    expanded.push(raw);
    for (const m of raw.matchAll(HTTP_URL_IN_TEXT)) {
      expanded.push(m[0]);
    }
  }
  return expanded;
}

function isSensitiveAdminBrowserPath(raw: string, toolName?: string): UrlGuardResult {
  if (!toolName || !PUPPETEER_TOOLS.has(toolName)) return { block: false };
  const parsed = parseUrlCandidate(normalizeUrlInput(raw));
  if (!parsed) return { block: false };
  const host = hostnameFromParsed(parsed).toLowerCase();
  if (DOCUMENTATION_HOST_ALLOWLIST.has(host)) return { block: false };
  const path = parsed.pathname || '/';
  for (const pattern of SENSITIVE_ADMIN_PATH_PATTERNS) {
    if (pattern.test(path)) {
      return {
        block: true,
        reason: `Blocked sensitive admin path for browser tool: ${path}`,
      };
    }
  }
  return { block: false };
}

export function evaluateUrlGuard(urls: string[], toolName?: string): UrlGuardResult {
  for (const raw of expandUrlCandidates(urls)) {
    const adminCheck = isSensitiveAdminBrowserPath(raw, toolName);
    if (adminCheck.block) return adminCheck;

    try {
      const parsed = parseUrlCandidate(normalizeUrlInput(raw));
      if (parsed) {
        const host = hostnameFromParsed(parsed).toLowerCase();
        if (isSpecDomainSquat(host)) {
          return {
            block: true,
            reason: `Blocked schema/json-schema subdomain squat: ${host}`,
          };
        }
      }
    } catch {
      /* fall through to other checks */
    }
    if (isTrustedDomainSquat(raw)) {
      return {
        block: true,
        reason: `Blocked trusted-domain subdomain squat: ${raw.slice(0, 80)}`,
      };
    }
    const check = isDangerousUrl(raw);
    if (check.block) {
      return { block: true, reason: check.reason ?? `Dangerous URL blocked: ${raw.slice(0, 80)}` };
    }
  }
  return { block: false };
}

File: adversarial-harness/python/policy_engine/url_guard.py
"""URL guard — mirrors url-guard.ts (SSRF / localhost / private IP)."""

from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Iterable
from urllib.parse import urlparse

from .arg_walker import walk_string_leaves

URL_ARG_FIELDS = frozenset(
    {"url", "href", "target", "webhook", "callback", "link", "message", "query", "body", "content", "text", "prompt"}
)
BLOCKED_SCHEMES = frozenset({"file", "javascript", "data", "vbscript", "about"})
DOCUMENTATION_HOST_ALLOWLIST = frozenset(
    {
        "example.com",
        "www.example.com",
        "docs.example.com",
    }
)
ALLOWED_SPEC_SCHEMA_HOSTS = frozenset(
    {
        "schema.org",
        "www.schema.org",
        "json-schema.org",
        "www.json-schema.org",
    }
)
SPEC_DOMAIN_SQUAT_RE = re.compile(r"^[\w-]+\.(?:schema|json-schema)\.org$", re.I)

LOCALHOST_NAMES = frozenset(
    {
        "localhost",
        "localhost.localdomain",
        "metadata",
        "metadata.google.internal",
        "metadata.google",
        "kubernetes.default.svc",
    }
)
PRIVATE_IPV4 = [
    re.compile(r"^127\."),
    re.compile(r"^10\."),
    re.compile(r"^192\.168\."),
    re.compile(r"^172\.(?:1[6-9]|2\d|3[01])\."),
    re.compile(r"^0\."),
    re.compile(r"^100\.(?:6[4-9]|[7-9]\d|1[01]\d|12[0-7])\."),
    re.compile(r"^169\.254\."),
]
METADATA_IPV4 = re.compile(r"^169\.254\.")
HTTP_URL_IN_TEXT = re.compile(r"https?://[^\s\"'<>]+", re.I)
PUPPETEER_TOOLS = frozenset({"puppeteer_navigate", "puppeteer_screenshot"})
SENSITIVE_ADMIN_PATH_PATTERNS = [
    re.compile(r"^/admin(?:/|$)", re.I),
    re.compile(r"^/wp-admin(?:/|$)", re.I),
    re.compile(r"^/wp-login\.php$", re.I),
    re.compile(r"^/dashboard(?:/|$)", re.I),
    re.compile(r"^/internal(?:/|$)", re.I),
    re.compile(r"^/management(?:/|$)", re.I),
    re.compile(r"^/console(?:/|$)", re.I),
    re.compile(r"^/settings(?:/|$)", re.I),
    re.compile(r"^/actuator(?:/|$)", re.I),
    re.compile(r"^/grafana(?:/|$)", re.I),
    re.compile(r"^/phpmyadmin(?:/|$)", re.I),
    re.compile(r"^/\.env$", re.I),
    re.compile(r"^/config(?:/|$)", re.I),
]


def _is_private_ipv4(host: str) -> bool:
    return any(p.search(host) for p in PRIVATE_IPV4)


def _decimal_ip_to_dotted(n: int) -> str:
    return ".".join(str((n >> shift) & 0xFF) for shift in (24, 16, 8, 0))


def _normalize_dotted_host(host: str) -> str:
    if not re.fullmatch(r"[\d.x]+", host, re.I) or "." not in host:
        return host
    parts = host.split(".")
    if len(parts) != 4:
        return host
    normalized: list[int] = []
    for part in parts:
        if re.fullmatch(r"0x[0-9a-f]+", part, re.I):
            normalized.append(int(part, 16) & 0xFF)
        elif re.fullmatch(r"0[0-7]+", part) and len(part) > 1:
            normalized.append(int(part, 8) & 0xFF)
        else:
            try:
                n = int(part, 10)
            except ValueError:
                return host
            if n < 0 or n > 255:
                return host
            normalized.append(n)
    return ".".join(str(n) for n in normalized)


def _is_decimal_ip_host(host: str) -> bool:
    if not re.fullmatch(r"\d{1,10}", host):
        return False
    n = int(host)
    return 0 <= n <= 0xFFFFFFFF and host == str(n)


def is_dangerous_url(raw: str) -> tuple[bool, str]:
    trimmed = raw.strip()
    if not trimmed:
        return False, ""
    if re.match(r"^(?:file|javascript|data|vbscript):", trimmed, re.I):
        return True, f"Blocked URL scheme: {trimmed[:32]}"

    try:
        parsed = urlparse(trimmed if re.match(r"^[a-z][a-z0-9+.-]*:", trimmed, re.I) else f"http://{trimmed}")
    except Exception:
        return False, ""

    scheme = (parsed.scheme or "").lower()
    if scheme in BLOCKED_SCHEMES:
        return True, f"Blocked URL scheme ({scheme})"

    host = (parsed.hostname or "").lower()
    if not host:
        return False, ""

    if re.fullmatch(r"0x[0-9a-f]+", host, re.I):
        try:
            n = int(host, 16)
            if 0 <= n <= 0xFFFFFFFF:
                host = _decimal_ip_to_dotted(n)
        except ValueError:
            pass
    elif re.fullmatch(r"[\d.x]+", host, re.I):
        host = _normalize_dotted_host(host)

    if host in LOCALHOST_NAMES or host.endswith(".localhost"):
        return True, f"Blocked localhost/metadata host: {host}"

    if _is_decimal_ip_host(host):
        dotted = _decimal_ip_to_dotted(int(host))
        if _is_private_ipv4(dotted) or METADATA_IPV4.search(dotted):
            return True, f"Blocked decimal IP (maps to {dotted})"

    if re.fullmatch(r"[\d.]+", host):
        if _is_private_ipv4(host) or METADATA_IPV4.search(host):
            return True, f"Blocked private/metadata IP: {host}"

    if ":" in host and (host == "::1" or host.startswith("fe80:") or host.startswith(("fc", "fd"))):
        return True, f"Blocked local/private IPv6: {host}"

    if METADATA_IPV4.search(host):
        return True, f"Blocked metadata IP: {host}"

    if host in DOCUMENTATION_HOST_ALLOWLIST:
        return False, ""

    return False, ""


def extract_http_urls_from_leaves(obj: object) -> list[str]:
    urls: list[str] = []
    for leaf in walk_string_leaves(obj):
        key = leaf.path.split(".")[-1].lower()
        if key in URL_ARG_FIELDS:
            urls.append(leaf.value)
        urls.extend(m.group(0) for m in HTTP_URL_IN_TEXT.finditer(leaf.value))
    return urls


@dataclass
class UrlGuardResult:
    block: bool
    reason: str = ""


def _is_spec_domain_squat(host: str) -> bool:
    h = host.lower()
    if h in ALLOWED_SPEC_SCHEMA_HOSTS:
        return False
    return bool(SPEC_DOMAIN_SQUAT_RE.match(h))


def _is_sensitive_admin_browser_path(raw: str, tool_name: str | None) -> UrlGuardResult:
    if not tool_name or tool_name not in PUPPETEER_TOOLS:
        return UrlGuardResult(block=False)
    try:
        parsed = urlparse(raw if "://" in raw else f"http://{raw}")
    except Exception:
        return UrlGuardResult(block=False)
    host = (parsed.hostname or "").lower()
    if host in DOCUMENTATION_HOST_ALLOWLIST:
        return UrlGuardResult(block=False)
    path = parsed.path or "/"
    for pat in SENSITIVE_ADMIN_PATH_PATTERNS:
        if pat.search(path):
            return UrlGuardResult(
                block=True,
                reason=f"Blocked sensitive admin path for browser tool: {path}",
            )
    return UrlGuardResult(block=False)


def evaluate_url_guard(urls: Iterable[str], tool_name: str | None = None) -> UrlGuardResult:
    expanded: list[str] = []
    for raw in urls:
        expanded.append(raw)
        expanded.extend(m.group(0) for m in HTTP_URL_IN_TEXT.finditer(raw))
    for raw in expanded:
        admin = _is_sensitive_admin_browser_path(raw, tool_name)
        if admin.block:
            return admin
        try:
            parsed = urlparse(raw if "://" in raw else f"http://{raw}")
            host = (parsed.hostname or "").lower()
            if host and _is_spec_domain_squat(host):
                return UrlGuardResult(
                    block=True,
                    reason=f"Blocked schema/json-schema subdomain squat: {host}",
                )
        except Exception:
            pass
        block, reason = is_dangerous_url(raw)
        if block:
            return UrlGuardResult(block=True, reason=reason)
    return UrlGuardResult(block=False)
